ggml.c

```diff
@@ -3550,16 +3550,17 @@ inline static void ggml_vec_log_f32  (const int n, float * y, const float * x) {
 inline static void ggml_vec_abs_f32  (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = fabsf(x[i]); }
 inline static void ggml_vec_sgn_f32  (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = (x[i] > 0.f) ? 1.f : ((x[i] < 0.f) ? -1.f : 0.f); }
 inline static void ggml_vec_step_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = (x[i] > 0.f) ? 1.f : 0.f; }
-inline static void ggml_vec_tanh_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = tanhf(x[i]);  }
+inline static void ggml_vec_tanh_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = tanh(x[i]);  }
 inline static void ggml_vec_elu_f32  (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = (x[i] > 0.f) ? x[i] : expf(x[i])-1; }
 inline static void ggml_vec_relu_f32 (const int n, float * y, const float * x) { for (int i = 0; i < n; ++i) y[i] = (x[i] > 0.f) ? x[i] : 0.f; }
 
-static const float GELU_COEF_A     = 0.044715f;
-static const float GELU_QUICK_COEF = -1.702f;
-static const float SQRT_2_OVER_PI  = 0.79788456080286535587989211986876f;
+static const float GELU_QUICK_COEF      = -1.702f;
+static const ggml_float GELU_COEF_A     = 0.044715;
+static const ggml_float SQRT_2_OVER_PI  = 0.79788456080286535587989211986876;
 
 inline static float ggml_gelu_f32(float x) {
-    return 0.5f*x*(1.0f + tanhf(SQRT_2_OVER_PI*x*(1.0f + GELU_COEF_A*x*x)));
+    const ggml_float xx = (ggml_float) x;
+    return 0.5*xx*(1.0 + tanh(SQRT_2_OVER_PI*xx*(1.0 + GELU_COEF_A*xx*xx)));
```

Comment:
How about this:
```
const float arg = SQRT_2_OVER_PI*x*(1.0 + GELU_COEF_A*x*x)
return arg < -10.f ? 0.f : arg < 10.f ? 0.5f*x*(1.0f + tanh(arg)) : x;
```
You are getting NaNs because of the absolute value of the argument of to the `tanh` function being too big `x`. If the hardware does not have a direct implementation of `tanh`, it will get evaluated as `tanh(x) = (exp(x) - exp(-x))/(exp(x) + exp(-x))`, so for sufficiently large arguments the `exp` will overflow, you will get `Inf/Inf`, which evaluates to `NaN`. `tanh( +/- 10) =  +/- 0.999999996`, so not evaluating outside of `+/- 10` is perfectly fine for single precision.

Reply:
> Oh, if you still keep getting `NaN` after implementing the function this way, it means you are getting `NaN` inputs.

Yes this is the case. Checking for `isnan(x)` in `ggml_gelu_f32` I found it even get NaNs as input when computing the lookup table here: This could be a bug in the FP16<-->F32 code?
https://github.com/ggerganov/llama.cpp/blob/01f2224682b08185af609b28b1268b95c8b4cfa2/ggml.c#L4331-L4332

Reply:
Probably those ranges should resolve to NaNs.